Java exceptions not handled by JDBCBackend

When installing ixmp from source into an otherwise clean new anaconda (python 3.7) environment outside the iiasa environment several test failed (see picture attached). Is this the expected behavior or did I do something wrong with the isntallation?


![image](https://user-images.githubusercontent.com/16190366/70852133-2acd8700-1e9e-11ea-83b8-9e521902dd37.png)

[pytest output](https://github.com/iiasa/ixmp/files/3968240/conda.prompt.test.report.txt)
<details>

```
====================================================== FAILURES =======================================================
____________________________________________________ test_init_set ____________________________________________________

test_mp = <ixmp.core.Platform object at 0x00000262BDD93488>

    def test_init_set(test_mp):
        """Test ixmp.Scenario.init_set()."""
        scen = ixmp.Scenario(test_mp, *can_args)

        # Add set on a locked scenario
        with pytest.raises(RuntimeError,
                           match="This Scenario cannot be edited, do a checkout "
                                 "first!"):
>           scen.init_set('foo')

tests\test_core.py:109:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\ixmp\core.py:756: in init_set
    return self._backend('init_item', 'set', name, idx_sets, idx_names)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\ixmp\core.py:358: in _backend
    return self.platform._backend(self, method, *args, **kwargs)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\ixmp\backend\base.py:25: in __call__
    return getattr(self, method)(obj, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <ixmp.backend.jdbc.JDBCBackend object at 0x00000262BDD93648>
s = <ixmp.core.Scenario object at 0x00000262BDD93788>, type = 'set', name = 'foo', idx_sets = None, idx_names = None

    def init_item(self, s, type, name, idx_sets, idx_names):
        # generate index-set and index-name lists
        if isinstance(idx_sets, set) or isinstance(idx_names, set):
            raise ValueError('index dimension must be string or ordered lists')
        idx_sets = to_jlist(idx_sets)
        idx_names = to_jlist(idx_names if idx_names is not None else idx_sets)

        # Initialize the Item
        func = getattr(self.jindex[s], f'initialize{type.title()}')

        # The constructor returns a reference to the Java Item, but these
        # aren't exposed by Backend, so don't return here
        try:
>           func(name, idx_sets, idx_names)
E           jpype._jclass.at.ac.iiasa.ixmp.exceptions.IxException: at.ac.iiasa.ixmp.exceptions.IxException: This Scenario cannot be edited, do a checkout first!

..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\ixmp\backend\jdbc.py:485: at.ac.iiasa.ixmp.exceptions.IxException
------------------------------------------------ Captured stdout setup ------------------------------------------------
2019-12-16 14:22:01,345 DEBUG at.ac.iiasa.ixmp.Platform:117 - Config: org.cfg4j.provider.BindInvocationHandler@64b8eb96
2019-12-16 14:22:01,346 DEBUG at.ac.iiasa.ixmp.database.DbDAO:176 - Performing DB migration for jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/init_set (user=ixmp, force=false)
2019-12-16 14:22:01,409 DEBUG at.ac.iiasa.ixmp.database.DbDAO:219 - hsqldb|Validate failed: Detected resolved migration not applied to database: 1.4
2019-12-16 14:22:01,503 DEBUG at.ac.iiasa.ixmp.database.DbDAO:2424 - URL (source):jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/init_set|ixmp|ixmp
2019-12-16 14:22:01,503  INFO at.ac.iiasa.ixmp.Platform:120 - Welcome to the IX modeling platform!
2019-12-16 14:22:01,504  INFO at.ac.iiasa.ixmp.Platform:121 -  connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/init_set' (user: ixmp)...
2019-12-16 14:22:01,506  INFO at.ac.iiasa.ixmp.Platform:129 - connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/init_set'...
------------------------------------------------- Captured log setup --------------------------------------------------
INFO     ixmp.backend.jdbc:jdbc.py:191 launching ixmp.Platform connected to jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/init_set
------------------------------------------------ Captured stdout call -------------------------------------------------
2019-12-16 14:22:01,508  INFO at.ac.iiasa.ixmp.objects.Scenario:209 - loading default Scenario 'canning problem|standard' (version: 2, runid: 2)...
2019-12-16 14:22:01,512  INFO at.ac.iiasa.ixmp.objects.Scenario:251 - done loading Scenario from the database!
2019-12-16 14:22:01,512  INFO at.ac.iiasa.ixmp.objects.Scenario:253 - Loaded: 0 sets, 4 parameters, 2 variables and 1 equations
_______________________________________________ test_reporter_visualize _______________________________________________

cmd = ['dot', '-Tpng']
input = b'digraph {\n\tgraph [rankdir=BT]\n\t-2787487202749238090 [label="" shape=box]\n\t-881712605417278162 [label="" shape=...268758 [label="" shape=box]\n\t6821383723090501221 [label="" shape=box]\n\t7074716014497090676 [label="" shape=box]\n}'
capture_output = True, check = True, quiet = False, kwargs = {'stderr': -1, 'stdin': -1, 'stdout': -1}

    def run(cmd, input=None, capture_output=False, check=False, quiet=False, **kwargs):
        """Run the command described by cmd and return its (stdout, stderr) tuple."""
        log.debug('run %r', cmd)
        if input is not None:
            kwargs['stdin'] = subprocess.PIPE
        if capture_output:
            kwargs['stdout'] = kwargs['stderr'] = subprocess.PIPE

        try:
>           proc = subprocess.Popen(cmd, startupinfo=get_startupinfo(), **kwargs)

..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\graphviz\backend.py:159:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <subprocess.Popen object at 0x00000262BDF4CF08>, args = ['dot', '-Tpng'], bufsize = -1, executable = None
stdin = -1, stdout = -1, stderr = -1, preexec_fn = None, close_fds = True, shell = False, cwd = None, env = None
universal_newlines = None, startupinfo = <subprocess.STARTUPINFO object at 0x00000262C12CF248>, creationflags = 0
restore_signals = True, start_new_session = False, pass_fds = ()

    def __init__(self, args, bufsize=-1, executable=None,
                 stdin=None, stdout=None, stderr=None,
                 preexec_fn=None, close_fds=True,
                 shell=False, cwd=None, env=None, universal_newlines=None,
                 startupinfo=None, creationflags=0,
                 restore_signals=True, start_new_session=False,
                 pass_fds=(), *, encoding=None, errors=None, text=None):
        """Create new Popen instance."""
        _cleanup()
        # Held while anything is calling waitpid before returncode has been
        # updated to prevent clobbering returncode if wait() or poll() are
        # called from multiple threads at once.  After acquiring the lock,
        # code must re-check self.returncode to see if another thread just
        # finished a waitpid() call.
        self._waitpid_lock = threading.Lock()

        self._input = None
        self._communication_started = False
        if bufsize is None:
            bufsize = -1  # Restore default
        if not isinstance(bufsize, int):
            raise TypeError("bufsize must be an integer")

        if _mswindows:
            if preexec_fn is not None:
                raise ValueError("preexec_fn is not supported on Windows "
                                 "platforms")
        else:
            # POSIX
            if pass_fds and not close_fds:
                warnings.warn("pass_fds overriding close_fds.", RuntimeWarning)
                close_fds = True
            if startupinfo is not None:
                raise ValueError("startupinfo is only supported on Windows "
                                 "platforms")
            if creationflags != 0:
                raise ValueError("creationflags is only supported on Windows "
                                 "platforms")

        self.args = args
        self.stdin = None
        self.stdout = None
        self.stderr = None
        self.pid = None
        self.returncode = None
        self.encoding = encoding
        self.errors = errors

        # Validate the combinations of text and universal_newlines
        if (text is not None and universal_newlines is not None
            and bool(universal_newlines) != bool(text)):
            raise SubprocessError('Cannot disambiguate when both text '
                                  'and universal_newlines are supplied but '
                                  'different. Pass one or the other.')

        # Input and output objects. The general principle is like
        # this:
        #
        # Parent                   Child
        # ------                   -----
        # p2cwrite   ---stdin--->  p2cread
        # c2pread    <--stdout---  c2pwrite
        # errread    <--stderr---  errwrite
        #
        # On POSIX, the child objects are file descriptors.  On
        # Windows, these are Windows file handles.  The parent objects
        # are file descriptors on both platforms.  The parent objects
        # are -1 when not using PIPEs. The child objects are -1
        # when not redirecting.

        (p2cread, p2cwrite,
         c2pread, c2pwrite,
         errread, errwrite) = self._get_handles(stdin, stdout, stderr)

        # We wrap OS handles *before* launching the child, otherwise a
        # quickly terminating child could make our fds unwrappable
        # (see #8458).

        if _mswindows:
            if p2cwrite != -1:
                p2cwrite = msvcrt.open_osfhandle(p2cwrite.Detach(), 0)
            if c2pread != -1:
                c2pread = msvcrt.open_osfhandle(c2pread.Detach(), 0)
            if errread != -1:
                errread = msvcrt.open_osfhandle(errread.Detach(), 0)

        self.text_mode = encoding or errors or text or universal_newlines

        # How long to resume waiting on a child after the first ^C.
        # There is no right value for this.  The purpose is to be polite
        # yet remain good for interactive users trying to exit a tool.
        self._sigint_wait_secs = 0.25  # 1/xkcd221.getRandomNumber()

        self._closed_child_pipe_fds = False

        try:
            if p2cwrite != -1:
                self.stdin = io.open(p2cwrite, 'wb', bufsize)
                if self.text_mode:
                    self.stdin = io.TextIOWrapper(self.stdin, write_through=True,
                            line_buffering=(bufsize == 1),
                            encoding=encoding, errors=errors)
            if c2pread != -1:
                self.stdout = io.open(c2pread, 'rb', bufsize)
                if self.text_mode:
                    self.stdout = io.TextIOWrapper(self.stdout,
                            encoding=encoding, errors=errors)
            if errread != -1:
                self.stderr = io.open(errread, 'rb', bufsize)
                if self.text_mode:
                    self.stderr = io.TextIOWrapper(self.stderr,
                            encoding=encoding, errors=errors)

            self._execute_child(args, executable, preexec_fn, close_fds,
                                pass_fds, cwd, env,
                                startupinfo, creationflags, shell,
                                p2cread, p2cwrite,
                                c2pread, c2pwrite,
                                errread, errwrite,
>                               restore_signals, start_new_session)

..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\subprocess.py:800:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <subprocess.Popen object at 0x00000262BDF4CF08>, args = 'dot -Tpng', executable = None, preexec_fn = None
close_fds = False, pass_fds = (), cwd = None, env = None
startupinfo = <subprocess.STARTUPINFO object at 0x00000262F3A03E48>, creationflags = 0, shell = False
p2cread = Handle(5016), p2cwrite = 87, c2pread = 88, c2pwrite = Handle(5980), errread = 89, errwrite = Handle(5932)
unused_restore_signals = True, unused_start_new_session = False

    def _execute_child(self, args, executable, preexec_fn, close_fds,
                       pass_fds, cwd, env,
                       startupinfo, creationflags, shell,
                       p2cread, p2cwrite,
                       c2pread, c2pwrite,
                       errread, errwrite,
                       unused_restore_signals, unused_start_new_session):
        """Execute program (MS Windows version)"""

        assert not pass_fds, "pass_fds not supported on Windows."

        if not isinstance(args, str):
            args = list2cmdline(args)

        # Process startup details
        if startupinfo is None:
            startupinfo = STARTUPINFO()
        else:
            # bpo-34044: Copy STARTUPINFO since it is modified above,
            # so the caller can reuse it multiple times.
            startupinfo = startupinfo._copy()

        use_std_handles = -1 not in (p2cread, c2pwrite, errwrite)
        if use_std_handles:
            startupinfo.dwFlags |= _winapi.STARTF_USESTDHANDLES
            startupinfo.hStdInput = p2cread
            startupinfo.hStdOutput = c2pwrite
            startupinfo.hStdError = errwrite

        attribute_list = startupinfo.lpAttributeList
        have_handle_list = bool(attribute_list and
                                "handle_list" in attribute_list and
                                attribute_list["handle_list"])

        # If we were given an handle_list or need to create one
        if have_handle_list or (use_std_handles and close_fds):
            if attribute_list is None:
                attribute_list = startupinfo.lpAttributeList = {}
            handle_list = attribute_list["handle_list"] = \
                list(attribute_list.get("handle_list", []))

            if use_std_handles:
                handle_list += [int(p2cread), int(c2pwrite), int(errwrite)]

            handle_list[:] = self._filter_handle_list(handle_list)

            if handle_list:
                if not close_fds:
                    warnings.warn("startupinfo.lpAttributeList['handle_list'] "
                                  "overriding close_fds", RuntimeWarning)

                # When using the handle_list we always request to inherit
                # handles but the only handles that will be inherited are
                # the ones in the handle_list
                close_fds = False

        if shell:
            startupinfo.dwFlags |= _winapi.STARTF_USESHOWWINDOW
            startupinfo.wShowWindow = _winapi.SW_HIDE
            comspec = os.environ.get("COMSPEC", "cmd.exe")
            args = '{} /c "{}"'.format (comspec, args)

        # Start the process
        try:
            hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                                     # no special security
                                     None, None,
                                     int(not close_fds),
                                     creationflags,
                                     env,
                                     os.fspath(cwd) if cwd is not None else None,
>                                    startupinfo)
E                                    FileNotFoundError: [WinError 2] Das System kann die angegebene Datei nicht finden

..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\subprocess.py:1207: FileNotFoundError

During handling of the above exception, another exception occurred:

test_mp = <ixmp.core.Platform object at 0x00000262BE1D6E08>
tmp_path = WindowsPath('C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/test_reporter_visualize0')

    def test_reporter_visualize(test_mp, tmp_path):
        scen = make_dantzig(test_mp)
        r = Reporter.from_scenario(scen)

>       r.visualize(str(tmp_path / 'visualize.png'))

tests\test_reporting.py:491:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\ixmp\reporting\__init__.py:559: in visualize
    return dask.visualize(self.graph, filename=filename, **kwargs)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\dask\base.py:530: in visualize
    return dot_graph(dsk, filename=filename, **kwargs)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\dask\dot.py:255: in dot_graph
    return graphviz_to_file(g, filename, format)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\dask\dot.py:267: in graphviz_to_file
    data = g.pipe(format=format)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\graphviz\files.py:138: in pipe
    quiet=quiet)
..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\graphviz\backend.py:229: in pipe
    out, _ = run(cmd, input=data, capture_output=True, check=True, quiet=quiet)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cmd = ['dot', '-Tpng']
input = b'digraph {\n\tgraph [rankdir=BT]\n\t-2787487202749238090 [label="" shape=box]\n\t-881712605417278162 [label="" shape=...268758 [label="" shape=box]\n\t6821383723090501221 [label="" shape=box]\n\t7074716014497090676 [label="" shape=box]\n}'
capture_output = True, check = True, quiet = False, kwargs = {'stderr': -1, 'stdin': -1, 'stdout': -1}

    def run(cmd, input=None, capture_output=False, check=False, quiet=False, **kwargs):
        """Run the command described by cmd and return its (stdout, stderr) tuple."""
        log.debug('run %r', cmd)
        if input is not None:
            kwargs['stdin'] = subprocess.PIPE
        if capture_output:
            kwargs['stdout'] = kwargs['stderr'] = subprocess.PIPE

        try:
            proc = subprocess.Popen(cmd, startupinfo=get_startupinfo(), **kwargs)
        except OSError as e:
            if e.errno == errno.ENOENT:
>               raise ExecutableNotFound(cmd)
E               graphviz.backend.ExecutableNotFound: failed to execute ['dot', '-Tpng'], make sure the Graphviz executables are on your systems' PATH

..\..\appdata\local\continuum\anaconda3\envs\ixmp\lib\site-packages\graphviz\backend.py:162: ExecutableNotFound
------------------------------------------------ Captured stdout setup ------------------------------------------------
2019-12-16 14:22:33,988 DEBUG at.ac.iiasa.ixmp.Platform:117 - Config: org.cfg4j.provider.BindInvocationHandler@5ce5e5e5
2019-12-16 14:22:33,990 DEBUG at.ac.iiasa.ixmp.database.DbDAO:176 - Performing DB migration for jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/reporter_visualize (user=ixmp, force=false)
2019-12-16 14:22:34,033 DEBUG at.ac.iiasa.ixmp.database.DbDAO:219 - hsqldb|Validate failed: Detected resolved migration not applied to database: 1.4
2019-12-16 14:22:34,259 DEBUG at.ac.iiasa.ixmp.database.DbDAO:2424 - URL (source):jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/reporter_visualize|ixmp|ixmp
2019-12-16 14:22:34,260  INFO at.ac.iiasa.ixmp.Platform:120 - Welcome to the IX modeling platform!
2019-12-16 14:22:34,261  INFO at.ac.iiasa.ixmp.Platform:121 -  connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/reporter_visualize' (user: ixmp)...
2019-12-16 14:22:34,262  INFO at.ac.iiasa.ixmp.Platform:129 - connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/reporter_visualize'...
------------------------------------------------- Captured log setup --------------------------------------------------
INFO     ixmp.backend.jdbc:jdbc.py:191 launching ixmp.Platform connected to jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/reporter_visualize
------------------------------------------------ Captured stdout call -------------------------------------------------
2019-12-16 14:22:34,277 DEBUG at.ac.iiasa.ixmp.objects.IndexSet:35 -  initialize index set 'i'
2019-12-16 14:22:34,278 DEBUG at.ac.iiasa.ixmp.objects.IndexSet:35 -  initialize index set 'j'
2019-12-16 14:22:34,279 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize parameter 'a' with explicitly stated index dimensions
2019-12-16 14:22:34,291 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize parameter 'b' with explicitly stated index dimensions
2019-12-16 14:22:34,298 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize parameter 'd' with explicitly stated index dimensions
2019-12-16 14:22:34,307 DEBUG at.ac.iiasa.ixmp.objects.Item:109 -  initialize parameter 'f' as scalar
2019-12-16 14:22:34,307 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize variable 'x' with explicitly stated index dimensions
2019-12-16 14:22:34,307 DEBUG at.ac.iiasa.ixmp.objects.Item:109 -  initialize variable 'z' as scalar
2019-12-16 14:22:34,307 DEBUG at.ac.iiasa.ixmp.objects.Item:109 -  initialize equation 'cost' as scalar
2019-12-16 14:22:34,308 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize equation 'demand' with explicitly stated index dimensions
2019-12-16 14:22:34,308 DEBUG at.ac.iiasa.ixmp.objects.Item:105 -  initialize equation 'supply' with explicitly stated index dimensions
2019-12-16 14:22:34,309  INFO at.ac.iiasa.ixmp.objects.Scenario:2106 - saving Scenario 'canning problem|standard' to database (runid: 3)...
2019-12-16 14:22:34,309 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2110 - run id: 3 - saved key-id map
2019-12-16 14:22:34,309 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2112 - run id: 3 - saved comments-id map
2019-12-16 14:22:34,311 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1569 - insert set 'i' rc=1 size=369
2019-12-16 14:22:34,312 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1569 - insert set 'j' rc=1 size=379
2019-12-16 14:22:34,313 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR a rc=1 size=412
2019-12-16 14:22:34,313 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR a rc=1 size=0
2019-12-16 14:22:34,313 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2137 - run id: 3 - saved parameter 'a'
2019-12-16 14:22:34,315 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR b rc=1 size=428
2019-12-16 14:22:34,315 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR b rc=1 size=0
2019-12-16 14:22:34,315 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2137 - run id: 3 - saved parameter 'b'
2019-12-16 14:22:34,316 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR d rc=1 size=472
2019-12-16 14:22:34,316 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR d rc=1 size=0
2019-12-16 14:22:34,316 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2137 - run id: 3 - saved parameter 'd'
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR f rc=1 size=390
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update PAR f rc=1 size=0
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2137 - run id: 3 - saved parameter 'f'
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update VAR x rc=1 size=0
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update VAR x rc=1 size=0
2019-12-16 14:22:34,318 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2157 - run id: 3 - saved variable 'x'
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update VAR z rc=1 size=0
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update VAR z rc=1 size=0
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2157 - run id: 3 - saved variable 'z'
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU cost rc=1 size=0
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU cost rc=1 size=0
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2171 - run id: 3 - saved equation 'cost'
2019-12-16 14:22:34,319 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU demand rc=1 size=0
2019-12-16 14:22:34,430 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU demand rc=1 size=0
2019-12-16 14:22:34,430 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2171 - run id: 3 - saved equation 'demand'
2019-12-16 14:22:34,650 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU supply rc=1 size=0
2019-12-16 14:22:34,872 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1927 - update EQU supply rc=1 size=0
2019-12-16 14:22:34,872 DEBUG at.ac.iiasa.ixmp.objects.Scenario:2171 - run id: 3 - saved equation 'supply'
2019-12-16 14:22:34,872 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:307 - Saving timeseries: added 0, removed: 0, updated 0
2019-12-16 14:22:34,979  INFO at.ac.iiasa.ixmp.objects.TimeSeries:988 - done saving Scenario 'canning problem|standard' to the database (runid: 3, version: 3)!
2019-12-16 14:22:34,980 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:135 - Cleaning cache...
2019-12-16 14:22:34,980 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:135 - Cleaning cache...
2019-12-16 14:22:34,981  INFO at.ac.iiasa.ixmp.objects.TimeSeries:1016 - assigned this Scenario as default version for 'canning problem|standard'
2019-12-16 14:22:34,989 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:234 - loading timeseries data for 0 variables...
2019-12-16 14:22:34,989 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:160 - Looking for cache by key TimeSeries.TsCache.CacheKey(key=GDP, unit=USD, node=DantzigLand, time=NOT_SPECIFIED, meta=1): 0 entries found
2019-12-16 14:22:34,997 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:234 - loading timeseries data for 0 variables...
2019-12-16 14:22:34,997 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:160 - Looking for cache by key TimeSeries.TsCache.CacheKey(key=Demand, unit=cases, node=DantzigLand, time=NOT_SPECIFIED, meta=0): 0 entries found
2019-12-16 14:22:34,997  INFO at.ac.iiasa.ixmp.objects.Scenario:2041 - committing changes of Scenario 'canning problem|standard' to the database (runid: 3)...
2019-12-16 14:22:35,000 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1555 - update set 'i' rc=1 size=369
2019-12-16 14:22:35,001 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1555 - update set 'j' rc=1 size=379
2019-12-16 14:22:35,003 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1555 - update set 'i' rc=1 size=369
2019-12-16 14:22:35,004 DEBUG at.ac.iiasa.ixmp.database.DbDAO:1555 - update set 'j' rc=1 size=379
2019-12-16 14:22:35,005 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:307 - Saving timeseries: added 2, removed: 0, updated 0
2019-12-16 14:22:35,224  INFO at.ac.iiasa.ixmp.objects.TimeSeries:985 - done updating Scenario 'canning problem|standard' to the database (runid: 3, version: 3)!
2019-12-16 14:22:35,224 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:135 - Cleaning cache...
2019-12-16 14:22:35,225 DEBUG at.ac.iiasa.ixmp.objects.TimeSeries:135 - Cleaning cache...
-------------------------------------------------- Captured log call --------------------------------------------------
DEBUG    graphviz.backend:backend.py:152 run ['dot', '-Tpng']
_____________________________________________________ test_close ______________________________________________________

test_mp = <ixmp.core.Platform object at 0x00000262C34B3748>
caplog = <_pytest.logging.LogCaptureFixture object at 0x00000262C34C2108>

    def test_close(test_mp, caplog):
        """Platform.close_db() doesn't throw needless exceptions."""
        # Close once
        test_mp.close_db()

        # Close again, once already closed
        test_mp.close_db()
>       assert caplog.records[0].message == \
            'Database connection could not be closed or was already closed'
E       AssertionError: assert 'at.ac.iiasa....e connection!' == 'Database con...lready closed'
E         - at.ac.iiasa.ixmp.exceptions.IxException: Error closing the database connection!
E         + Database connection could not be closed or was already closed

..\..\tests\backend\test_jdbc.py:35: AssertionError
------------------------------------------------ Captured stdout setup ------------------------------------------------
2019-12-16 14:22:58,780 DEBUG at.ac.iiasa.ixmp.Platform:117 - Config: org.cfg4j.provider.BindInvocationHandler@9005147
2019-12-16 14:22:58,781 DEBUG at.ac.iiasa.ixmp.database.DbDAO:176 - Performing DB migration for jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close (user=ixmp, force=false)
2019-12-16 14:22:58,826 DEBUG at.ac.iiasa.ixmp.database.DbDAO:219 - hsqldb|Validate failed: Detected resolved migration not applied to database: 1.4
2019-12-16 14:22:58,872 DEBUG at.ac.iiasa.ixmp.database.DbDAO:2424 - URL (source):jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close|ixmp|ixmp
2019-12-16 14:22:58,872  INFO at.ac.iiasa.ixmp.Platform:120 - Welcome to the IX modeling platform!
2019-12-16 14:22:58,873  INFO at.ac.iiasa.ixmp.Platform:121 -  connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close' (user: ixmp)...
2019-12-16 14:22:58,873  INFO at.ac.iiasa.ixmp.Platform:129 - connected to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close'...
------------------------------------------------- Captured log setup --------------------------------------------------
INFO     ixmp.backend.jdbc:jdbc.py:191 launching ixmp.Platform connected to jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close
------------------------------------------------ Captured stdout call -------------------------------------------------
2019-12-16 14:22:59,224  INFO at.ac.iiasa.ixmp.Platform:137 - closed the connection to database 'jdbc:hsqldb:file:C:/Users/ga46gup/AppData/Local/Temp/pytest-of-ga46gup/pytest-3/localdb/close'
2019-12-16 14:22:59,224 ERROR at.ac.iiasa.ixmp.database.DbDAO:1013 - Error closing the database connection!
java.sql.SQLException: HikariDataSource HikariDataSource (HikariPool-83) has been closed.
        at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:96)
        at at.ac.iiasa.ixmp.database.HsqldbDAO.shutdownIfNeeded(HsqldbDAO.java:30)
        at at.ac.iiasa.ixmp.database.DbDAO.closeConn(DbDAO.java:2452)
        at at.ac.iiasa.ixmp.Platform.closeDB(Platform.java:136)
-------------------------------------------------- Captured log call --------------------------------------------------
WARNING  ixmp.backend.jdbc:jdbc.py:247 at.ac.iiasa.ixmp.exceptions.IxException: Error closing the database connection!
```
</details>

> I have attached the output from the last 5000 lines.
> 
> [conda prompt test report.txt](https://github.com/iiasa/ixmp/files/3968240/conda.prompt.test.report.txt)

Thank you! I pulled out the relevant lines and put them in the issue description above.

Respectively:
1. test_init_set: In this case, the code in JDBCBackend should catch the exception and re-raise it as a RuntimeError, as show here: https://github.com/iiasa/ixmp/blob/399e1b04db32b8fb6f090dba611a900fd4c421ea/ixmp/backend/jdbc.py#L484-L489 I'm not sure why this is failing in your environment.
2. test_reporter_visualize:
   ```
   graphviz.backend.ExecutableNotFound: failed to execute ['dot', '-Tpng'], make sure the Graphviz executables are on your systems' PATH
   ```
   This tests a reporting feature that requires the graphviz *program*. ixmp automated installation of the python-graphviz package, but you probably will need to install this program.
3. test_close:
   ```
   AssertionError: assert 'at.ac.iiasa....e connection!' == 'Database con...lready closed'
     - at.ac.iiasa.ixmp.exceptions.IxException: Error closing the database connection!
     + Database connection could not be closed or was already closed
   ```
   This seems similar to (1).

At first glance, what's happening in (1) and (3) is that the exceptions returned by the Java/ixmp_source code, when converted to `str()`, have the class name prepended (`"at.ac.iiasa.ixmp.exceptions.IxException: This Scenario cannot..."` instead of `"This Scenario cannot..."`). ixmp does not expect this, so it fails to convert these Java exceptions into normal Python exceptions.

I've changed the title to better describe the issue. We will need to think a little further to diagnose why this happens on your system but not on Travis, AppVeyor, or other users' systems here.

The good news is that ixmp is functioning properly; the only issue is that some code might raise a different exception class from `RuntimeError`. If you are not using `try: ... except:` clauses in your code, this is probably not an issue.

--- a/RELEASE_NOTES.rst
+++ b/RELEASE_NOTES.rst
@@ -4,6 +4,7 @@ Next release
 All changes
 -----------
 
+- :pull:`362`: Raise Python exceptions from :class:`.JDBCBackend`.
 - :pull:`354`: Add :meth:`Scenario.items`, :func:`.utils.diff`, and allow using filters in CLI command ``ixmp export``.
 - :pull:`353`: Add meta functionality.
 

--- a/ixmp/backend/jdbc.py
+++ b/ixmp/backend/jdbc.py
@@ -264,13 +264,22 @@ class JDBCBackend(CachingBackend):
         except jpype.JException as e:  # pragma: no cover
             # Handle Java exceptions
             jclass = e.__class__.__name__
-            if jclass.endswith('HikariPool.PoolInitializationException'):
+            if jclass.endswith("HikariPool.PoolInitializationException"):
                 redacted = copy(kwargs)
-                redacted.update({'user': '(HIDDEN)', 'password': '(HIDDEN)'})
-                msg = f'unable to connect to database:\n{repr(redacted)}'
-            elif jclass.endswith('FlywayException'):
-                msg = 'when initializing database:'
-            _raise_jexception(e, f'{msg}\n(Java: {jclass})')
+                redacted.update({"user": "(HIDDEN)", "password": "(HIDDEN)"})
+                msg = f"unable to connect to database:\n{repr(redacted)}"
+            elif jclass.endswith("FlywayException"):
+                msg = "when initializing database:"
+                if "applied migration" in e.args[0]:
+                    msg += (
+                        "\n\nThe schema of the database does not match the "
+                        "schema of this version of ixmp. To resolve, either "
+                        "install the version of ixmp used to create the "
+                        "database, or delete it and retry."
+                    )
+            else:
+                _raise_jexception(e)
+            raise RuntimeError(f"{msg}\n(Java: {jclass})")
 
         # Set the log level
         self.set_log_level(log_level)
@@ -334,12 +343,8 @@ class JDBCBackend(CachingBackend):
         """
         try:
             self.jobj.closeDB()
-        except java.IxException as e:
-            if 'Error closing the database connection!' in str(e):
-                log.warning('Database connection could not be closed or was '
-                            'already closed')
-            else:
-                log.warning(str(e))
+        except java.IxException as e:  # pragma: no cover
+            log.warning(str(e))
         except AttributeError:
             pass  # self.jobj is None, e.g. cleanup after __init__ fails
 
@@ -621,7 +626,10 @@ class JDBCBackend(CachingBackend):
         try:
             self.jindex[ts].commit(comment)
         except java.IxException as e:
-            _raise_jexception(e)
+            if "this Scenario is not checked out" in e.args[0]:
+                raise RuntimeError(e.args[0])
+            else:  # pragma: no cover
+                _raise_jexception(e)
         if ts.version == 0:
             ts.version = self.jindex[ts].getVersion()
 
@@ -930,10 +938,14 @@ class JDBCBackend(CachingBackend):
                 # - par: (value, unit, comment)
                 jobj.addElement(*args)
         except java.IxException as e:
-            msg = e.message()
-            if ('does not have an element' in msg) or ('The unit' in msg):
+            if any(
+                s in e.args[0] for s in
+                ("does not have an element", "The unit")
+            ):
                 # Re-raise as Python ValueError
-                raise ValueError(msg) from None
+                raise ValueError(e.args[0]) from None
+            elif "cannot be edited" in e.args[0]:
+                raise RuntimeError(e.args[0])
             else:  # pragma: no cover
                 _raise_jexception(e)
 

--- a/ixmp/tests/backend/test_jdbc.py
+++ b/ixmp/tests/backend/test_jdbc.py
@@ -61,6 +61,16 @@ def test_close(test_mp, capfd):
     assert msg in captured.out
 
 
+def test_exceptions(test_mp):
+    """Ensure that Python exceptions are raised for some actions."""
+    s = ixmp.Scenario(test_mp, "model name", "scenario name", "new")
+    s.init_par("foo", [])
+    s.commit("")
+
+    with pytest.raises(RuntimeError):
+        s.change_scalar("foo", 42, unit="kg")
+
+
 def test_pass_properties():
     ixmp.Platform(driver='hsqldb', url='jdbc:hsqldb:mem://ixmptest',
                   user='ixmp', password='ixmp')

